[CMake] Add additional check for libatomic

dpino opened the pull request:
#### 7ccde68358e76d232faf0cb79cc464c746c780aa
<pre>
[CMake] Add additional check for libatomic
<a href="https://bugs.webkit.org/show_bug.cgi?id=292267">https://bugs.webkit.org/show_bug.cgi?id=292267</a>

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

In 32-bit systems, atomic data types larger than 8-bytes require to link with
&apos;libatomic&apos;.

* Source/cmake/WebKitCompilerFlags.cmake: Add additional check for determining
  whether to require libatomic.

Canonical link: <a href="https://commits.webkit.org/294274@main">https://commits.webkit.org/294274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2a0a39ee141166835250f22d77d29ba294256dd7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101318 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11284 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106468 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51945 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21289 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29475 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77166 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34203 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104325 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16419 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91512 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57513 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16237 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51293 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93985 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86111 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9598 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108822 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99927 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28446 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20922 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86142 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28808 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87719 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85701 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30425 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8133 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22551 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16486 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28376 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123553 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28188 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34372 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29746 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -458,6 +458,28 @@ int main() {
+
+int main(int argc, char* argv[])
+{
+    std::atomic<std::optional<double>> d;
```
We already have an `ATOMICS_REQUIRE_LIBATOMIC` variable defined in CMake. We should change the check that calculates it if needed, instead of adding a new one.

dpino replied:
I prefer not to mix up this program with the existing one, because it will make the code less clearer. I think it's better to have different programs for different scenarios.

aperezdc replied:
Thing is, the previous program should have _already_ worked to detect whether libatomic is needed or not. So it needs fixing.

aperezdc replied:
Anyway, I believe we do not need this for now, given that the regression was solved in https://github.com/WebKit/WebKit/pull/44682 by switching to `float` instead of using `double`, so I guess we can close this PR for now and revisit the issue when needed later on. Right?

dpino replied:
No, we don't want to do that. It may happen again in the future that someone uses an atomic data type larger than 8-bytes and the build will break again in 32-bit. The check that determines whether to link with libatomic has to be fixed.

dpino replied:
I merged this additonal check into the existing program.

aperezdc reviewed (approved):
Much nicer now, r=me.

dpino committed:
[CMake] Add additional check for libatomic
https://bugs.webkit.org/show_bug.cgi?id=292267

Reviewed by Michael Catanzaro and Adrian Perez de Castro.

In 32-bit systems, atomic data types larger than 8-bytes require to link with
'libatomic'.

* Source/cmake/WebKitCompilerFlags.cmake: Add additional check for determining
  whether to require libatomic.

Canonical link: https://commits.webkit.org/294274@main
